### PR TITLE
Fast functions

### LINKED ISSUE #26: Consider dropping T_JULIA_FUNC and instead using T_FUNCTION objects with custom handlers
Consider these timings, where `myfunc` is a simple Julia function that simply returns its first argument:
```
gap> ListX([1..10^4], [1..2], {i,j} -> i);; time;
14
gap> ListX([1..10^4], [1..2], myfunc);; time;
595
```
So Julia is not competitive at all for function calls.

I think this could be improved by using `T_FUNCTION` objects to wrap Julia functions; this avoids various slow paths in calling the current `T_JULIA_FUNC` objects.

To do this, you need to store the pointer to the julia function somewhere in the `T_FUNCTION`. I think `FEXS_FUNC` should be OK for that. Also set `BODY_FUNC` to zero, so that GAP considers the `T_FUNCTION` object as a kernel handler.

Then rewrite `Func_JuliaCallFunc0Arg` etc. into `DoJuliaCallFunc0Arg` handlers, to be hooked into the various `HDLR_FUNC` slots. I could work on this if you are interested, but I thought I should first find out if there was a reason you haven't used this approach so far.

### PR DESCRIPTION
**Pull request by fingolfin**
Resolves #26

Consider this Julia file:
```julia
module Stuff

import Main.ForeignGAP: MPtr

function myfunc( a::MPtr, b::MPtr )
    return a;
end

function myfunc( a::MPtr, b::Int64 )
    return a;
end

function myfunc( a::Int64, b::MPtr )
    return a;
end

function myfunc( a::Int64, b::Int64 )
    return a;
end

end

function blafunc(a, b)
    return a
end
```

Then on the GAP side, I load them using this (`_JuliaFunctionByModuleNoConv` and `_JuliaFunctionNoConv` are added by this PR):
```gap
JuliaIncludeFile( "stuff.jl" );
if not IsBound(_JuliaFunctionByModuleNoConv) then
  _JuliaFunctionByModuleNoConv:= _JuliaFunctionByModule;
  _JuliaFunctionNoConv:= _JuliaFunction;
fi;

myfunc := JuliaFunction("myfunc", "Stuff");
myfuncNoConv := _JuliaFunctionByModuleNoConv("myfunc", "Stuff");
myfuncC := JuliaBindCFunction( "Stuff.myfunc", 2, ["a", "b"] );

blafunc := JuliaFunction("blafunc");
blafuncNoConv := _JuliaFunctionNoConv("blafunc");
blafuncC := JuliaBindCFunction( "blafunc", 2, ["a", "b"] );
```

Then run some benchmarks, comparing them against a native GAP function, as well as the kernel function `ReturnFirst`

Before this PR:
```
gap> GASMAN("collect"); ListX([1..10^5], [1..10], {i,j} -> i);; time;
365
gap> GASMAN("collect"); ListX([1..10^5], [1..10], ReturnFirst);; time;
296
gap> GASMAN("collect"); ListX([1..10^5], [1..10], myfunc);; time;
2821
gap> GASMAN("collect"); ListX([1..10^5], [1..10], blafunc);; time;
3275
gap> GASMAN("collect"); ListX([1..10^5], [1..10], myfuncC);; time;
1803
gap> GASMAN("collect"); ListX([1..10^5], [1..10], blafuncC);; time;
1695
gap>
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], {i,j} -> i);; time;
293
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], ReturnFirst);; time;
288
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], myfunc);; time;
1888
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], blafunc);; time;
2047
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], myfuncC);; time;
1679
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], blafuncC);; time;
1578
gap>
gap> input:=ListWithIdenticalEntries(10^5,fail);;
gap> GASMAN("collect"); ListX("0123456789", input, {i,j} -> i);; time;
371
gap> GASMAN("collect"); ListX("0123456789", input, ReturnFirst);; time;
268
gap> GASMAN("collect"); ListX("0123456789", input, myfunc);; time;
1844
gap> GASMAN("collect"); ListX("0123456789", input, blafunc);; time;
1954
gap> GASMAN("collect"); ListX("0123456789", input, myfuncC);; time;
1447
gap> GASMAN("collect"); ListX("0123456789", input, blafuncC);; time;
1483
```

With this PR:
```
gap> GASMAN("collect"); ListX([1..10^5], [1..10], {i,j} -> i);; time;
252
gap> GASMAN("collect"); ListX([1..10^5], [1..10], ReturnFirst);; time;
271
gap> GASMAN("collect"); ListX([1..10^5], [1..10], myfunc);; time;
816
gap> GASMAN("collect"); ListX([1..10^5], [1..10], blafunc);; time;
863
gap> GASMAN("collect"); ListX([1..10^5], [1..10], myfuncC);; time;
1685
gap> GASMAN("collect"); ListX([1..10^5], [1..10], blafuncC);; time;
1748
gap> GASMAN("collect"); ListX([1..10^5], [1..10], myfuncNoConv);; time;
869
gap> GASMAN("collect"); ListX([1..10^5], [1..10], blafuncNoConv);; time;
873
gap>
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], {i,j} -> i);; time;
285
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], ReturnFirst);; time;
261
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], myfunc);; time;
284
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], blafunc);; time;
299
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], myfuncC);; time;
1559
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], blafuncC);; time;
1626
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], myfuncNoConv);; time;
278
gap> GASMAN("collect"); ListX("0123456789", [1..10^5], blafuncNoConv);; time;
277
gap>
gap> input:=ListWithIdenticalEntries(10^5,fail);;
gap> GASMAN("collect"); ListX("0123456789", input, {i,j} -> i);; time;
357
gap> GASMAN("collect"); ListX("0123456789", input, ReturnFirst);; time;
271
gap> GASMAN("collect"); ListX("0123456789", input, myfunc);; time;
272
gap> GASMAN("collect"); ListX("0123456789", input, blafunc);; time;
283
gap> GASMAN("collect"); ListX("0123456789", input, myfuncC);; time;
1494
gap> GASMAN("collect"); ListX("0123456789", input, blafuncC);; time;
1489
gap> GASMAN("collect"); ListX("0123456789", input, myfuncNoConv);; time;
249
gap> GASMAN("collect"); ListX("0123456789", input, blafuncNoConv);; time;
258
```

Some observations:
* The GAP kernel function `ReturnFirst` is sometimes actually slower than the native GAP function `{i,j} -> i`. I suspect that is because `ReturnFirst` is variadic, so its arguments are first packed into a plist and then extracted again.
* With this PR, the function wrappers created by `JuliaFunction` are *faster* than those created by `JuliaBindCFunction`. However, I plan to optimize the latter as well later on, then I hope they'll catch up again.
* The `NoConv` functions are not much (if at all) faster then the ones which perform automatic conversion of arguments. Since immediate integers always have to be converted some way, I also added the test where some of the arguments definitely will not be converted. These then *are* noticably faster, and indeed, on par with the GAP kernel function resp. the native GAP function.

In conclusion, it seems that at least in this micro-benchmark, there seems to be considerable overhead creating a Julia `int64` boxing GAP immediate integers.

### CODE REVIEW

**Comment by sebasguts** [comment]
Is this finished now, and should I review and ultimately merge it, or is it still WIP?

**Comment by fingolfin** [comment]
It can be reviewed, but I should fix the commit messages; and the second commit perhaps should be revised (but that shube discussed)
